### PR TITLE
Add support for sidekiq_retry_in and sidekiq_retries_exhausted_block for ActiveJob

### PR DESCRIPTION
**Pull request by pierrickouw**
Hello 👋 

Trying to make the `sidekiq_retries_exhausted` and `sidekiq_retry_in` works with ActiveJob.
There some issue describing this behavior: https://github.com/sidekiq/sidekiq/issues/4496, and some workaround existing (for `sidekiq_retries_exhausted`), but none (as far as I know) for `sidekiq_retry_in`.

It is also written in the Sidekiq Wiki as currently not supported: https://github.com/sidekiq/sidekiq/wiki/Active-Job#customizing-error-handling, but it would be a huge bonus for us

Open to suggestions, other idea I had was to modify the [ActiveJob Adapter](https://github.com/rails/rails/blob/main/activejob/lib/active_job/queue_adapters/sidekiq_adapter.rb) to keep it in sync and modify the Job Instance on the fly, but thought it would be better to keep the current support as much as possible inside Sidekiq. [(That what we did as a workaround in the meantime)](https://gist.github.com/pierrickouw/da5fb6bffda2b17ab50d427ade1beb83)

There might be better places aswell in Sidekiq codebase. 

Thanks!

### CODE REVIEW

**Comment by mperham** [comment]
IMO if you are using ActiveJobs, you should be using its retry subsystem and semantics. If you want Sidekiq::Job's retry subsystem, use Sidekiq::Job.

**Comment by pierrickouw** [comment]
Thanks for your answer!

I kind of agree, but as there is already a support for simple things like `sidekiq_options`, it often make people confused on why it doesn't work (and it is silent) when they are going through the documentation. As AJ is the default in Rails, many people goes this way aswell (I understand that Sidekiq is not intended _for_ Rails). The usage is a bit different (like the global identifier), so the migration from one system to the other is not straightforward (and addition like this one would certainly helps people move to a full Sidekiq implementation and benefits the performance upgrade once they are ready, cause the retry will already have the proper mechanism and semantics - _at least it would help us for sure, that's why we did the monkey patch_)

Retry system in AJ is very basic, and does not have the richness of Sidekiq's one, so it helps bridge the gap. (but no problem if you think it doesn't bring any values / is not aligned with Sidekiq philosophy!)

**Comment by mperham** [comment]
I'm reconsidering this. Can you add a test for the exhausted block, because you have a bug as is. 😃

**Comment by pierrickouw** [comment]
Ok sure, will do tomorrow! Thanks!

**Comment by pierrickouw** [comment]
@mperham Fixed it and added a test!

File: test/retry_exhausted_test.rb [hunk 1]
```diff
@@ -55,6 +64,10 @@ def old_worker
     @old_worker ||= OldJob.new
   end
 
+  def wrapped_worker
```
Comment (mperham) [inline test/retry_exhausted_test.rb:67]: Do you need this?

File: test/retry_test.rb [hunk 2]
```diff
@@ -381,6 +387,12 @@ def job
         end
         assert_equal 1, ds.size
       end
+
+      it "supports wrapped jobs" do
+        strat, count = handler.__send__(:delay_for, ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper, 2, StandardError.new, {"wrapped" => "ActiveJobRetry"})
```
Comment (mperham) [inline test/retry_test.rb:392]: Use `ActiveJobRetry.to_s` instead so the code works if we refactor the class name.

**Review by mperham — CHANGES_REQUESTED** [review]
Minor nits. Almost there!

**Comment by pierrickouw** [comment]
Should be good, thanks!

File: test/retry_exhausted_test.rb (hunk 1 above)
Comment (pierrickouw) [reply test/retry_exhausted_test.rb]: Not really, can remove it

**Review by mperham — APPROVED** [review]

**Comment by mperham** [comment]
Thank you, will be available in 7.1.3.

**Pull request merged**